python/ray/serve/deployment.py

```diff
@@ -120,6 +120,17 @@ def __init__(
         self._replica_config = replica_config
         self._docs_path = docs_path
 
+    def _validate_name(self, name: str):
+        if not isinstance(name, str):
+            raise TypeError("name must be a string.")
+
+        # name does not contain #
+        if "#" in name:
+            warnings.warn(
+                f"Deployment names cannot contain '#' character, this will raise an error in a future release. "
```

Comment:
It's probably easier for users if we give them a specific release. Let's do 2 or 3 releases from now?

Also
```suggestion
                f"Deployment names should not contain the '#' character, this will raise an error in a future release. "
```